File: python/1d-dynamic-programming/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        cache = {}

        def dfs(current):
            if current > amount:
                return float('inf')
            if current == amount:
                return 0
            if current in cache:
                return cache[current]

            cache[current] = float('inf')
            for coin in coins:
                cache[current] = min(cache[current], dfs(current + coin) + 1)

            return cache[current]

        answer = dfs(0)
        if answer < float('inf'):
            return answer

        return -1

File: c++/1d-dynamic-programming/coin-change.cc
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        std::vector<int> count(amount + 1, std::numeric_limits<int>::max());
        count[0] = 0;
        for (int coin : coins) {
            for (int i = coin; i < amount + 1; ++i) {
                if (count[i - coin] < std::numeric_limits<int>::max()) {
                    count[i] = std::min(count[i], count[i - coin] + 1);
                }
            }
        }
        
        if (count[amount] == std::numeric_limits<int>::max()) {
            return -1;
        }

        return count[amount];
    }
};
